### DISCUSSION TITLE
oauth2 with svelte kit server side callback

### DISCUSSION THREAD

**Discussion by mingsterism** [pocketbase/pocketbase#4514] [Q&A]
hi guys, 
i'm trying to get google login to work with the urlCallback but can't seem to get it to work
```
export const actions: Actions = {
    loginGoogle: async ({request, locals}) => {
        let callbackUrl = ""
        const resultUrl = await locals.pb.collection('users').authWithOAuth2({
                provider: 'google',
                urlCallback: (url) => {
                    callbackUrl = url           // the url here is correct from google
                    return url
                },

            }
        )
        return {
            err: null,
            data: {
                url: callbackUrl
            }
        }
    },
 }
 ```
 this is my code in +page.server.ts
 
 however, i'm unable to do any redirects in the actions. i'm also unable to return back the url to the client side. 
 is there any way to handle this in server side svelte? 

thanks

**Comment by ganigeorgiev** [comment]
This is a very frequent question and was asked at least several times before but I really don't know how to explain it better as it seems that most users don't have a clear understanding how the OAuth2 works, where their code execute and why they need to complicate their flow so much by using JS SSR in the first place (see some ramblings related to this in https://github.com/pocketbase/pocketbase/discussions/4412, https://github.com/pocketbase/pocketbase/discussions/3792#discussioncomment-7620032 and probably others).

In short, the `authWithOAuth2` call can't be executed from a node server action because it needs to initialize a long-lived SSE connection with the PocketBase server and at the same time it needs to open a browser client-side window.

If you want to handle the OAuth2 flow on the server-side you'll need to implement the "Manual code exchange" flow (_the second tab in [OAuth2 integration](https://pocketbase.io/docs/authentication/#oauth2-integration)_). The docs show a plain JS browser example, but the same idea could be applied in a node server context - you practically need 2 routes:
1. Endpoint that will show the "Login with ..." button(s) (this usually is done with the help of `listAuthMethods`).
2. Endpoint that will handle the OAuth2 redirect and will call `authWithOAuth2Code` (aka. the actual PocketBase authentication)
The main difference from the example in the docs would be that you'll have to replace the localStorage with cookies so that the redirect handler can access the `codeVerifier` and `state` initialized from step1

Note that when using the "Manual code exchange" flow the OAuth2 app redirect url must be the endpoint from step2.
